File: providers/aws/src/main/java/cloudspec/aws/ec2/resource/nested/EC2InstanceNetworkInterface.java
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * #L%
 */
package cloudspec.aws.ec2.resource.nested;

import cloudspec.annotation.AssociationDefinition;
import cloudspec.annotation.PropertyDefinition;
import cloudspec.aws.ec2.resource.EC2NetworkInterfaceResource;
import cloudspec.aws.ec2.resource.EC2SecurityGroupResource;
import cloudspec.aws.ec2.resource.EC2SubnetResource;
import cloudspec.aws.ec2.resource.EC2VpcResource;
import software.amazon.awssdk.services.ec2.model.GroupIdentifier;
import software.amazon.awssdk.services.ec2.model.InstanceIpv6Address;
import software.amazon.awssdk.services.ec2.model.InstanceNetworkInterface;

import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;

public class EC2InstanceNetworkInterface {
    @AssociationDefinition(
            name = "groups",
            description = "One or more security groups",
            targetClass = EC2SecurityGroupResource.class
    )
    private final List<String> groupIds;

    @PropertyDefinition(
            name = "ipv6_addresses",
            description = "One or more IPv6 addresses associated with the network interface"
    )
    private final List<String> ipv6Addresses;

    @AssociationDefinition(
            name = "network_interface",
            description = "The network interface",
            targetClass = EC2NetworkInterfaceResource.class
    )
    private final String networkInterfaceId;

    @PropertyDefinition(
            name = "owner_id",
            description = "The ID of the AWS account that created the network interface"
    )
    private final String ownerId;

    @PropertyDefinition(
            name = "private_dns_name",
            description = "The private DNS name"
    )
    private final String privateDnsName;

    @PropertyDefinition(
            name = "private_ip_address",
            description = "The IPv4 address of the network interface within the subnet"
    )
    private final String privateIpAddress;

    @PropertyDefinition(
            name = "private_ip_addresses",
            description = "One or more private IPv4 addresses associated with the network interface"
    )
    private final List<EC2InstancePrivateIpAddress> privateIpAddresses;

    @PropertyDefinition(
            name = "source_dest_check",
            description = "Indicates whether to validate network traffic to or from this network interface"
    )
    private final Boolean sourceDestCheck;

    @PropertyDefinition(
            name = "status",
            description = "The status of the network interface",
            exampleValues = "available | associated | attached | in-use | detaching"
    )
    private final String status;

    @AssociationDefinition(
            name = "subnet",
            description = "The subnet",
            targetClass = EC2SubnetResource.class
    )
    private final String subnetId;

    @AssociationDefinition(
            name = "vpc",
            description = "The VPC",
            targetClass = EC2VpcResource.class
    )
    private final String vpcId;

    @PropertyDefinition(
            name = "interface_type",
            description = "Describes the type of network interface",
            exampleValues = "interface | efa"
    )
    private final String interfaceType;

    public EC2InstanceNetworkInterface(List<String> groupIds, List<String> ipv6Addresses,
                                       String networkInterfaceId, String ownerId, String privateDnsName,
                                       String privateIpAddress, List<EC2InstancePrivateIpAddress> privateIpAddresses,
                                       Boolean sourceDestCheck, String status, String subnetId, String vpcId,
                                       String interfaceType) {
        this.groupIds = groupIds;
        this.ipv6Addresses = ipv6Addresses;
        this.networkInterfaceId = networkInterfaceId;
        this.ownerId = ownerId;
        this.privateDnsName = privateDnsName;
        this.privateIpAddress = privateIpAddress;
        this.privateIpAddresses = privateIpAddresses;
        this.sourceDestCheck = sourceDestCheck;
        this.status = status;
        this.subnetId = subnetId;
        this.vpcId = vpcId;
        this.interfaceType = interfaceType;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        EC2InstanceNetworkInterface that = (EC2InstanceNetworkInterface) o;
        return Objects.equals(groupIds, that.groupIds) &&
                Objects.equals(ipv6Addresses, that.ipv6Addresses) &&
                Objects.equals(networkInterfaceId, that.networkInterfaceId) &&
                Objects.equals(ownerId, that.ownerId) &&
                Objects.equals(privateDnsName, that.privateDnsName) &&
                Objects.equals(privateIpAddress, that.privateIpAddress) &&
                Objects.equals(privateIpAddresses, that.privateIpAddresses) &&
                Objects.equals(sourceDestCheck, that.sourceDestCheck) &&
                Objects.equals(status, that.status) &&
                Objects.equals(subnetId, that.subnetId) &&
                Objects.equals(vpcId, that.vpcId) &&
                Objects.equals(interfaceType, that.interfaceType);
    }

    @Override
    public int hashCode() {
        return Objects.hash(groupIds, ipv6Addresses, networkInterfaceId, ownerId, privateDnsName,
                privateIpAddress, privateIpAddresses, sourceDestCheck, status, subnetId, vpcId, interfaceType);
    }

    public static List<EC2InstanceNetworkInterface> fromSdk(List<InstanceNetworkInterface> instanceNetworkInterfaces) {
        return Optional.ofNullable(instanceNetworkInterfaces)
                .orElse(Collections.emptyList())
                .stream()
                .map(EC2InstanceNetworkInterface::fromSdk)
                .collect(Collectors.toList());
    }

    public static EC2InstanceNetworkInterface fromSdk(InstanceNetworkInterface instanceNetworkInterface) {
        return new EC2InstanceNetworkInterface(
                securityGroupIdsFromSdk(instanceNetworkInterface.groups()),
                ipv6AddressesFromSdk(instanceNetworkInterface.ipv6Addresses()),
                instanceNetworkInterface.networkInterfaceId(),
                instanceNetworkInterface.ownerId(),
                instanceNetworkInterface.privateDnsName(),
                instanceNetworkInterface.privateIpAddress(),
                EC2InstancePrivateIpAddress.fromSdk(instanceNetworkInterface.privateIpAddresses()),
                instanceNetworkInterface.sourceDestCheck(),
                instanceNetworkInterface.statusAsString(),
                instanceNetworkInterface.subnetId(),
                instanceNetworkInterface.vpcId(),
                instanceNetworkInterface.interfaceType()
        );
    }

    public static List<String> securityGroupIdsFromSdk(List<GroupIdentifier> groupIdentifiers) {
        return Optional.ofNullable(groupIdentifiers)
                .orElse(Collections.emptyList())
                .stream()
                .map(GroupIdentifier::groupId)
                .collect(Collectors.toList());
    }

    public static List<String> ipv6AddressesFromSdk(List<InstanceIpv6Address> instanceIpv6Addresses) {
        return Optional.ofNullable(instanceIpv6Addresses)
                .orElse(Collections.emptyList())
                .stream()
                .map(InstanceIpv6Address::ipv6Address)
                .collect(Collectors.toList());
    }

}

File: providers/aws/src/main/kotlin/cloudspec/aws/ec2/nested/EC2InstanceNetworkInterface.kt
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package cloudspec.aws.ec2.nested

import cloudspec.annotation.AssociationDefinition
import cloudspec.annotation.PropertyDefinition
import cloudspec.aws.ec2.EC2NetworkInterface
import cloudspec.aws.ec2.EC2SecurityGroup
import cloudspec.aws.ec2.EC2Subnet
import cloudspec.aws.ec2.EC2Vpc

data class EC2InstanceNetworkInterface(
        @property:AssociationDefinition(
                name = ASSOC_GROUPS,
                description = ASSOC_GROUPS_D,
                targetClass = EC2SecurityGroup::class
        )
        val groupIds: List<String>?,

        @property:PropertyDefinition(
                name = PROP_IPV6_ADDRESSES,
                description = PROP_IPV6_ADDRESSES_D
        )
        val ipv6Addresses: List<String>?,

        @property:AssociationDefinition(
                name = ASSOC_NETWORK_INTERFACE,
                description = ASSOC_NETWORK_INTERFACE_D,
                targetClass = EC2NetworkInterface::class
        )
        val networkInterfaceId: String?,

        @property:PropertyDefinition(
                name = PROP_OWNER_ID,
                description = PROP_OWNER_ID_D
        )
        val ownerId: String?,

        @property:PropertyDefinition(
                name = PROP_PRIVATE_DNS_NAME,
                description = PROP_PRIVATE_DNS_NAME_D
        )
        val privateDnsName: String?,

        @property:PropertyDefinition(
                name = PROP_PRIVATE_IP_ADDRESS,
                description = PROP_PRIVATE_IP_ADDRESS_D
        )
        val privateIpAddress: String?,

        @property:PropertyDefinition(
                name = PROP_PRIVATE_IP_ADDRESSES,
                description = PROP_PRIVATE_IP_ADDRESSES_D
        )
        val privateIpAddresses: List<EC2InstancePrivateIpAddress>?,

        @property:PropertyDefinition(
                name = PROP_SOURCE_DEST_CHECK,
                description = PROP_SOURCE_DEST_CHECK_D
        )
        val sourceDestCheck: Boolean?,

        @property:PropertyDefinition(
                name = PROP_STATUS,
                description = PROP_STATUS_D,
                exampleValues = "available | associated | attached | in-use | detaching"
        )
        val status: String?,

        @property:AssociationDefinition(
                name = ASSOC_SUBNET,
                description = ASSOC_SUBNET_D,
                targetClass = EC2Subnet::class
        )
        val subnetId: String?,

        @property:AssociationDefinition(
                name = ASSOC_VPC,
                description = ASSOC_VPC_D,
                targetClass = EC2Vpc::class
        )
        val vpcId: String?,

        @property:PropertyDefinition(
                name = PROP_INTERFACE_TYPE,
                description = PROP_INTERFACE_TYPE_D,
                exampleValues = "interface | efa"
        )
        val interfaceType: String?
) {
    companion object {
        const val ASSOC_GROUPS = "groups"
        const val ASSOC_GROUPS_D = "One or more security groups"
        const val PROP_IPV6_ADDRESSES = "ipv6_addresses"
        const val PROP_IPV6_ADDRESSES_D = "One or more IPv6 addresses associated with the network interface"
        const val ASSOC_NETWORK_INTERFACE = "network_interface"
        const val ASSOC_NETWORK_INTERFACE_D = "The network interface"
        const val PROP_OWNER_ID = "owner_id"
        const val PROP_OWNER_ID_D = "The ID of the AWS account that created the network interface"
        const val PROP_PRIVATE_DNS_NAME = "private_dns_name"
        const val PROP_PRIVATE_DNS_NAME_D = "The private DNS name"
        const val PROP_PRIVATE_IP_ADDRESS = "private_ip_address"
        const val PROP_PRIVATE_IP_ADDRESS_D = "The IPv4 address of the network interface within the subnet"
        const val PROP_PRIVATE_IP_ADDRESSES = "private_ip_addresses"
        const val PROP_PRIVATE_IP_ADDRESSES_D = "One or more private IPv4 addresses associated with the network interface"
        const val PROP_SOURCE_DEST_CHECK = "source_dest_check"
        const val PROP_SOURCE_DEST_CHECK_D = "Indicates whether to validate network traffic to or from this network interface"
        const val PROP_STATUS = "status"
        const val PROP_STATUS_D = "The status of the network interface"
        const val ASSOC_SUBNET = "subnet"
        const val ASSOC_SUBNET_D = "The subnet"
        const val ASSOC_VPC = "vpc"
        const val ASSOC_VPC_D = "The VPC"
        const val PROP_INTERFACE_TYPE = "interface_type"
        const val PROP_INTERFACE_TYPE_D = "Describes the type of network interface"
    }
}
